### PR TITLE
Add a flag for rcon login via username

### PR DESCRIPTION
**Pull request by Learath2**
Added a new flag in serverinfo to allow checking whether a server requires username logins. No longer need to use the empty username.

### CODE REVIEW

**Comment by heinrich5991** [comment]
Now this sounds like a perfect use-case for the new extended protocol.

**Comment by Learath2** [comment]
Well this works too doesn't it? I couldn't really think of an advantage in using a message.

**Comment by heinrich5991** [comment]
The advantage of using a message would be that we don't collide with other servers using that flag.

**Comment by Learath2** [comment]
Thought it was unused, I'll go ahead and make it a message.

**Comment by Learath2** [comment]
I either have to send a game message from `engine/server.cpp` or need to add an exception to pass a system message to `CGameClient::OnMessage` so it can get to the components. This really feels like it belongs to serverinfo, maybe a new serverinfo request?

**Comment by heinrich5991** [comment]
@Learath2 Or you can mirror the way the rcon lines from the server get passed through to the console component.

Server info is inadequate for this in my opinion, server info is for stuff that is interesting *before* you join the server.

File: src/engine/server/authmanager.h [hunk 1]
```diff
@@ -46,6 +46,7 @@ class CAuthManager
 	void ListKeys(FListCallback pfnListCallbac, void *pUser);
 	void AddDefaultKey(int Level, const char *pPw);
 	bool IsGenerated();
+	int NonDefaultKeys();
```
Comment (heinrich5991) [inline src/engine/server/authmanager.h]: Maybe `NumNonDefaultKeys`?

File: src/engine/server/server.cpp [hunk 2]
```diff
@@ -864,6 +864,13 @@ int CServer::DelClientCallback(int ClientID, const char *pReason, void *pUser)
 	return 0;
 }
 
+void CServer::SendRconType(int ClientID, bool UsernameReq)
+{
+	CMsgPacker Msg(NETMSG_RCONTYPE);
+	Msg.AddInt(UsernameReq);
+	SendMsgEx(&Msg, MSGFLAG_VITAL|MSGFLAG_FLUSH, ClientID, true);
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: Probably shouldn't be `MSGFLAG_FLUSH`, at least for the initial send (it'll get sent when `NETMSG_MAP_CHANGE` is sent).

File: src/engine/client/client.cpp [hunk 3]
```diff
@@ -1957,6 +1957,11 @@ void CClient::ProcessServerPacket(CNetChunk *pPacket)
 				}
 			}
 		}
+		else if(Msg == NETMSG_RCONTYPE)
+		{
+			bool UsernameReq = Unpacker.GetInt();
```
Comment (heinrich5991) [inline src/engine/client/client.cpp]: Maybe do `Unpacker.GetInt() & 1` for future-compatiblity, should we decide to add another flag.

File: src/engine/server/authmanager.h (hunk 1 above)
Comment (Learath2) [inline src/engine/server/authmanager.h]: That's what I had originally, felt too verbose. Can do tho.

File: src/engine/server/server.cpp (hunk 2 above)
Comment (Learath2) [inline src/engine/server/server.cpp]: I guess it never needs to be flushed instantly.

**Comment by Learath2** [comment]
Failed to `git commit` :P

**Comment by heinrich5991** [comment]
Reopened at #800.

**Pull request closed without merging**